# PD01 · Explorar y preparar AGROSAVIA

[Índice y distribución de las cuatro horas](../ejercicios/README.md) · [Guía de pandas](../docs/manual-pandas.md)

**Tiempo de trabajo: 60 minutos.** El objetivo es reconocer la estructura de un DataFrame y preparar medidas y fechas conservando los valores originales.

| Símbolo | Uso |
|:---:|---|
| 📘 | Concepto o función de apoyo. |
| ✅ | Evidencia que debe quedar visible. |
| ⚠️ | Límite de interpretación. |

Esta plantilla se encuentra en `ejercicios/PD01_exploracion_agrosavia.ipynb`. Antes de comenzar, se crea una copia en `soluciones/PD01_exploracion_agrosavia.ipynb` dentro del repositorio de cada estudiante. El desarrollo y la entrega se realizan en esa copia; la plantilla de `ejercicios/` se conserva como enunciado. En JupyterLab se puede duplicar el archivo y mover la copia a `soluciones/`, manteniendo el nombre indicado; si ya existe una solución, se continúa sobre ella. Las instrucciones se presentan en celdas Markdown; las celdas de código marcadas como **Desarrollo** contienen el espacio de trabajo. Cada actividad incluye una celda para explicar el resultado.

Se utiliza el kernel del entorno `.venv` del curso. Los CSV ya deben estar descargados mediante `kit/00_datos.py`. La celda inicial prepara las rutas y carga únicamente las entradas de este ejercicio; el desarrollo del análisis corresponde al estudiante. Este notebook se ejecuta de forma independiente.

**Apoyo en el manual:** secciones 1–5 y la receta de registros en revisión de la sección 11.

## Ubicación de trabajo: la raíz del repositorio

📘 **La raíz del repositorio** es la carpeta principal de la copia del proyecto. En la instalación del curso se llama `big-data-postgraduate` y contiene `README.md`, `kit/`, `ejercicios/` y `soluciones/`:

```text
bigdata/
└── big-data-postgraduate/       ← raíz del repositorio
    ├── README.md
    ├── kit/
    ├── ejercicios/              ← enunciados y plantillas
    └── soluciones/              ← notebooks resueltos
```

**Desde la terminal de Ubuntu en WSL.** `pwd` muestra la carpeta actual, `cd` cambia de carpeta y `ls` muestra su contenido. Los siguientes comandos se ejecutan en la terminal, no en una celda Python:

```bash
pwd
cd /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate
pwd
ls
```

✅ Después de `cd`, el segundo `pwd` debe mostrar `/mnt/c/Users/TUPTC/bigdata/big-data-postgraduate`. En la salida de `ls` deben aparecer `README.md`, `kit`, `ejercicios` y `soluciones`. La carpeta `bigdata` es la carpeta contenedora; la raíz del proyecto es su subcarpeta `big-data-postgraduate`.

La ruta anterior corresponde a la configuración de la guía WSL. Si el repositorio se clonó en otra ubicación, se sustituye por la ruta real de esa copia. Si `cd` informa que la carpeta no existe, se revisa esa ubicación antes de continuar.

**Desde una subcarpeta conocida.** `..` significa «carpeta superior». Se comprueba primero la ubicación con `pwd` y se utiliza la alternativa correspondiente:

| Ubicación actual, relativa al repositorio | Comando para volver a la raíz |
|---|---|
| `ejercicios/`, `soluciones/` o `kit/` | `cd ..` |
| `kit/Notebooks/` | `cd ../..` |

Al llegar a la raíz, se puede activar el entorno y abrir JupyterLab desde la terminal:

```bash
source .venv/bin/activate
jupyter lab
```

**Dentro de un notebook ya abierto.** La terminal y el kernel de Python mantienen sus propios directorios de trabajo. La celda de carga siguiente localiza la raíz y la guarda en `REPO`, aunque el notebook esté abierto desde `soluciones/` o `ejercicios/`. Después de ejecutar esa celda, el siguiente bloque puede ejecutarse en una celda Python para comprobar la ubicación y cambiar el directorio del kernel a la raíz:

```python
import os
from pathlib import Path

print("Directorio antes del cambio:", Path.cwd())
print("Raíz detectada:", REPO)
os.chdir(REPO)
print("Directorio después del cambio:", Path.cwd())
```

✅ `Path.cwd()` debe coincidir con `REPO` después del cambio. Esta operación cambia la base para interpretar rutas relativas; el archivo del notebook permanece en `soluciones/`. La carga de datos de estos ejercicios también funciona sin ese cambio porque construye las rutas a partir de `REPO`.

## Preparación

Se ejecuta la siguiente celda y se comprueba que el intérprete pertenece al entorno del curso y que pandas corresponde a la versión 2.2.3. Las rutas se calculan desde el repositorio, sin introducir rutas personales de Windows, WSL o macOS.

In [1]:
from pathlib import Path
import sys
import pandas as pd
from IPython.display import display

actual = Path.cwd().resolve()
REPO = next(
    (ruta for ruta in [actual, *actual.parents]
     if (ruta / "kit/fuentes.json").is_file()
     and (ruta / "kit/00_datos.py").is_file()),
    None,
)
if REPO is None:
    raise FileNotFoundError("El notebook debe abrirse dentro del repositorio del curso")

RAW = REPO / "kit/data/raw"
OUT = REPO / "kit/salidas/soluciones/PD01"
archivos = ['agrosavia.csv']
faltantes = [nombre for nombre in archivos if not (RAW / nombre).is_file()]
if faltantes:
    raise FileNotFoundError(f"Faltan CSV en {RAW}: {faltantes}")
OUT.mkdir(parents=True, exist_ok=True)

opciones = dict(sep=",", encoding="utf-8-sig", dtype=str, keep_default_na=False)
suelo = pd.read_csv(RAW / "agrosavia.csv", **opciones)


print("Python:", sys.version.split()[0])
print("Intérprete:", sys.executable)
print("pandas:", pd.__version__)
print("Repositorio:", REPO)
print("Salidas:", OUT)

Python: 3.12.15
Intérprete: /mnt/c/Users/TUPTC/.venv/bin/python
pandas: 2.2.3
Repositorio: /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate
Salidas: /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate/kit/salidas/soluciones/PD01


## 1. Inspección inicial · 10 minutos

Presentar la cantidad de filas y columnas, los nombres de las columnas, sus tipos y cinco registros. Explicar qué representa una fila y por qué la lectura inicial conserva los campos como texto.

📘 **Apoyo:** `shape`, `columns`, `dtypes`, `head` e `info`.

✅ **Comprobación:** El corte del manifiesto contiene 92.738 filas y 32 columnas. La tabla original debe conservar esa cantidad de filas durante el ejercicio.

In [2]:
import pandas as pd
# Desarrollo de la actividad 1
# Mostrar dimensiones, tipos y primeros registros de suelo.
ds_agrosavia = pd.read_csv('../kit/data/raw/agrosavia.csv', keep_default_na=False)
# Guardar el total de filas original para los controles posteriores.
ds_agrosavia.info()
ds_agrosavia.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92738 entries, 0 to 92737
Data columns (total 32 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Secuencial                          92738 non-null  int64 
 1   Fecha de Análisis                   92738 non-null  object
 2   Departamento                        92738 non-null  object
 3   Municipio                           92738 non-null  object
 4   Cultivo                             92738 non-null  object
 5   Estado                              92738 non-null  object
 6   Tiempo de establecimiento           92738 non-null  object
 7   Topografia                          92738 non-null  object
 8   Drenaje                             92738 non-null  object
 9   Riego                               92738 non-null  object
 10  Fertilizantes aplicados             92738 non-null  object
 11  pH agua:suelo                       92738 non-null  ob

/tmp/ipykernel_685/1085375329.py:4: DtypeWarning: Columns (11,12,13,14,17,18,20,21) have mixed types. Specify dtype option on import or set low_memory=False.
  ds_agrosavia = pd.read_csv('../kit/data/raw/agrosavia.csv', keep_default_na=False)


,Secuencial,Fecha de Análisis,Departamento,Municipio,Cultivo,Estado,Tiempo de establecimiento,Topografia,Drenaje,Riego,...,Conductividad electrica,Hierro disponible olsen,Cobre disponible,Manganeso disponible Olsen,Zinc disponible Olsen,Boro disponible,Hierro disponible doble acido,Cobre disponible doble acido,Manganeso disponible doble acido,Zinc disponible doble acido
0,1,7/01/2014,NARIÑO,SAN ANDRÉS DE TUMACO,No Indica,No indica,No indica,No indica,No indica,No indica,...,0.133,66.39,1.700,1.700,1.5,0.187,ND,ND,ND,ND
1,2,21/08/2014,HUILA,SANTA MARÍA,Granadilla,Establecido,De 1 a 5 años,Pendiente,Buen drenaje,No Tiene,...,2.749,250.9,3.7,29.4,28.70,2.084,ND,ND,ND,ND
2,3,22/08/2014,ANTIOQUIA,LIBORINA,Café,Por establecer,No indica,Pendiente,Mal drenaje,No Tiene,...,0.328,390,11.10,8.299,5,0.085,ND,ND,ND,ND
3,4,22/08/2014,ANTIOQUIA,LIBORINA,Maracuyá,Por establecer,No indica,Ondulado,Mal drenaje,No Tiene,...,0.171,200,4.2,4.699,1,0.229,ND,ND,ND,ND
4,5,22/08/2014,ANTIOQUIA,LIBORINA,Café,Por establecer,No indica,Ondulado,Mal drenaje,No Tiene,...,0.323,117.9,3.8,4.600,0.700,0.139,ND,ND,ND,ND


**Interpretación del resultado:**

Pues se pudo comprobar que hay 92.738 filas y 32 columnas en el archivo de agrosavia.csv. Con la consulta anterior, se pudo saber que a excepción de la primera columna que es de typo "int64", el resto de columnas es de tipo "object".Tambien se vio el nombre de cada columna y los primeros 5 registros del archivo, cada registro es una fila que representa los datos de una produccion de un cultivo agricola en determinado municipio, que ademas hace parte de un departamento. 

Por otro lado la lectura inicial conserva los campos como texto, porque esta contando cuantas filas tiene el archivo, como se llama cada columna, la cantidad de filas nulas y el tipo de informacion que recibe cada columna.

## 2. Diagnóstico de cinco campos · 10 minutos

Construir una tabla llamada `perfil` para `Secuencial`, `Fecha de Análisis`, `Departamento`, `Municipio` y `pH agua:suelo`. Para cada campo se registran los nulos de pandas, los textos vacíos después de retirar espacios y los marcadores `ND`. Las tres situaciones se cuentan por separado.

📘 **Apoyo:** `isna`, `.str.strip`, `.str.upper`, `.eq` y `sum`.

✅ **Comprobación:** `perfil` debe tener una fila por campo. Con `keep_default_na=False`, un texto vacío o `ND` no se convierte automáticamente en un nulo.

In [3]:
#Desarrollo de la actividad 2
# Crear y mostrar perfil con las cinco columnas solicitadas.

import pandas as pd
campos = ['Secuencial', 'Fecha de Análisis', 'Departamento', 'Municipio', 'pH agua:suelo']
registros = []

for campo in campos:
    s = ds_agrosavia[campo]
    
    # Nulos nativos de pandas (NaN / None)
    nulos = s.isna().sum()
    
    # Cadenas no nulas con espacios removidos
    s_str = s.dropna().astype(str).str.strip()
    
    # Textos vacíos tras retirar espacios
    textos_vacios = s_str.eq("").sum()
    
    # Marcadores "ND" 
    marcadores_nd = s_str.str.upper().eq("ND").sum()
    
    registros.append({
        "campo": campo,
        "nulos": nulos,
        "textos_vacios": textos_vacios,
        "marcadores_ND": marcadores_nd
    })

# 2. Construir el DataFrame 'perfil' final
perfil = pd.DataFrame(registros)

display(perfil)


,campo,nulos,textos_vacios,marcadores_ND
0,Secuencial,0,0,0
1,Fecha de Análisis,0,0,0
2,Departamento,0,0,0
3,Municipio,0,2,0
4,pH agua:suelo,0,0,0


**Interpretación del resultado:**

Despues de contruir la tabla llamada "perfil", con las columnas Secuencial, Fecha de Análisis, Departamento, Municipio y pH agua:suelo. Se realizo el conteo por separado de los valores nulos, los espacios vacios y los marcadores ND, donde se pudo observar que en la tabla "perfil", solo hay en la columna "Municipios" dos lugares con texto vacio, de resto en la tabla no hay valores nulos, espacios vacios o marcadores ND. Como aclaracion en el punto 1, se le añadio el comando "keep_default_na=False", para que no se tomara automaticamente un espacio vacio o un valor ND, como nulo, porque en caso de que no se colocara se tomaria como valores nulos, los dos espacios vacios que estan en la tabla.

## 3. Conversión de pH y fecha · 15 minutos

Crear `ph` desde `pH agua:suelo` y `fecha_dt` desde `Fecha de Análisis`. Utilizar conversión numérica con `errors="coerce"` y fecha con formato `%d/%m/%Y`. Conservar los originales. Distinguir entradas vacías de textos no vacíos que no pudieron convertirse.

📘 **Apoyo:** `pd.to_numeric`, `.str.replace`, `pd.to_datetime`, `isna` y máscaras booleanas.

✅ **Comprobación:** Se muestran los conteos de conversiones fallidas y hasta cinco ejemplos con su valor original. No se reemplazan medidas ausentes por cero.

In [4]:
# Desarrollo de la actividad 3
# Crear suelo["ph"] y suelo["fecha_dt"].
# Mostrar los fallos de conversión y sus valores originales.
ds_agrosavia["ph"] = pd.to_numeric(
    ds_agrosavia["pH agua:suelo"].str.strip().str.replace(",", ".", regex=False),
    errors="coerce",
)
ds_agrosavia["fecha_dt"] = pd.to_datetime(
    ds_agrosavia["Fecha de Análisis"],
    format="%d/%m/%Y",
    errors="coerce",
)
#suelo["anio"] = suelo["fecha_dt"].dt.year.astype("Int64")

display(ds_agrosavia[["ph","fecha_dt"]])


,ph,fecha_dt
0,5.64,2014-01-07
1,4.24,2014-08-21
2,5.22,2014-08-22
3,5.45,2014-08-22
4,5.43,2014-08-22
...,...,...
92733,NaN,2025-04-07
92734,NaN,2025-04-04
92735,NaN,2025-04-03
92736,NaN,2025-04-04


In [5]:
vacia = ds_agrosavia["pH agua:suelo"].str.strip().eq("")
vacia1 = ds_agrosavia["Fecha de Análisis"].str.strip().eq("")
no_numerica = ~vacia & ds_agrosavia["ph"].isna()
no_numerica1 = ~vacia & ds_agrosavia["fecha_dt"].isna()

print("Vacías_ph:", int(vacia.sum()))
print("Vacías_fecha:", int(vacia1.sum()))
print("No numéricas:", int(no_numerica.sum()))
print(ds_agrosavia.loc[no_numerica, ["Secuencial", "pH agua:suelo"]].head())
print("---------------")
print("No numéricas:", int(no_numerica1.sum()))
print(ds_agrosavia.loc[no_numerica1, ["Secuencial", "Fecha de Análisis"]].head())

Vacías_ph: 0
Vacías_fecha: 0
No numéricas: 27213
       Secuencial pH agua:suelo
7282         7283         4..01
16384       16385          5.65
16385       16386          5.21
16386       16387          5.31
16387       16388           4.9
---------------
No numéricas: 0
Empty DataFrame
Columns: [Secuencial, Fecha de Análisis]
Index: []


**Interpretación del resultado:**

Se creo la tabla "suelo" con las columnas "ph" y "fecha_dt", como se vio anteriormente.Tambien se utilizo la conversion numerica y la fecha con formato "%d/%m/%Y". Por otro lado, se observo que no hay vacios ni en la columna 'ph', ni en la columna 'Fecha_dt' pero si se encontro textos no vacios que no se pudo convertir en la columna "ph", por mala redaccion o uso de simbolos como se ve anteriormente. En la columna "fecha_dt" no aparecio texto que no se convirtio.

## 4. Registros para revisión · 15 minutos

Crear la columna booleana `en_revision`. Se marca una fila si su pH es ausente, está fuera de `[0, 14]` o su fecha no se pudo interpretar. Contar por separado cada motivo y el total de filas marcadas. Conservar todas las observaciones.

📘 **Apoyo:** `between`, `isna`, `|`, `~`, `.loc` y `value_counts`.

✅ **Comprobación:** Las filas marcadas y no marcadas deben sumar el total original. Una misma fila puede tener varios motivos; por eso la suma de los motivos no necesariamente coincide con el total marcado.

In [6]:
# Desarrollo de la actividad 4
# Crear en_revision y los conteos por motivo.
# Comprobar la conservación del total original con assert.
import pandas as pd

en_revision = ds_agrosavia.dropna()
# 1. Identificación individual de cada motivo
motivo_ph_ausente = ds_agrosavia['ph'].isna()
motivo_ph_fuera_rango = ~ds_agrosavia['ph'].between(0, 14) & ds_agrosavia['ph'].notna()
motivo_fecha_invalida = ds_agrosavia['fecha_dt'].isna()  # Suponiendo que las fechas no interpretables son NaT/NaN

# 2. Crear en_revision y los conteos por motivo
en_revision = motivo_ph_ausente | motivo_ph_fuera_rango | motivo_fecha_invalida

# Conteos por motivo separado (usando .loc para cumplir con el apoyo metodológico)
conteo_ph_ausente = len(ds_agrosavia.loc[motivo_ph_ausente])
conteo_ph_fuera_rango = len(ds_agrosavia.loc[motivo_ph_fuera_rango])
conteo_fecha_invalida = len(ds_agrosavia.loc[motivo_fecha_invalida])

# Resumen de motivos
conteos_motivos = pd.Series({
    'pH ausente': conteo_ph_ausente,
    'pH fuera de rango [0, 14]': conteo_ph_fuera_rango,
    'Fecha no interpretable': conteo_fecha_invalida
})

# Distribución general de filas marcadas vs. no marcadas
resumen_revision = en_revision.value_counts()

# 3. Comprobar la conservación del total original con assert
total_original = len(ds_agrosavia)
total_evaluado = resumen_revision.sum()

assert total_evaluado == total_original, (
    f"Error de integridad: La suma de categorías ({total_evaluado}) "
    f"difiere del total de filas original ({total_original})."
)

# Imprimir resultados
display('en_revision')
print("--- Conteos por Motivo ---")
print(conteos_motivos)
print("\n--- Distribución de 'en_revision' ---")
print(resumen_revision)
print(f"\n Validación 'assert' superada: {total_evaluado} filas evaluadas de {total_original} originales.")

'en_revision'

--- Conteos por Motivo ---
pH ausente                   27213
pH fuera de rango [0, 14]        0
Fecha no interpretable           0
dtype: int64

--- Distribución de 'en_revision' ---
False    65525
True     27213
Name: count, dtype: int64

 Validación 'assert' superada: 92738 filas evaluadas de 92738 originales.


**Interpretación del resultado:**

Se puede observar que se creo la columna "en_revision", donde se marcaba una fila si su pH es ausente,si su pH esta fuera de [0,14] o su fecha no se pudo interpretar. Pero revisando los resultados, se encontro que existen 27213 filas que cumplen la condicion que el valor de pH esta ausente, mientra que no se presebtaron alguna fila que cumpliera con la condicion de que el pH este fuera de 0 a 14, ni que una fecha no se pueda interpretar. Por otro lado, las filas marcada y las no marcadas, si suma el valor original de filas que tenia pH en ds_agrosavia.

## 5. Evidencias y conclusiones · 10 minutos

Construir `controles` con las columnas `control` y `cantidad`: total de entrada, pH sin número, pH fuera del intervalo, fecha no interpretable y total de filas en revisión. Exportar a `OUT / "controles_agrosavia.csv"` sin índice. Redactar dos hallazgos apoyados en los conteos.

📘 **Apoyo:** `pd.DataFrame`, `to_csv(index=False)` y las tablas de las actividades anteriores.

✅ **Comprobación:** El archivo se genera dentro de la ruta `OUT`. Cada hallazgo menciona un conteo y explica su significado o limitación.

In [7]:
# Desarrollo de la actividad 5
# Construir controles, mostrarlo y exportarlo a OUT.
# Redactar los dos hallazgos en la celda Markdown siguiente.
from pathlib import Path
OUT = Path("OUT")
OUT.mkdir(exist_ok=True)
controles = pd.DataFrame({
    "control": [
        "total de entrada",
        "pH sin número",
        "pH fuera del intervalo",
        "fecha no interpretable",
        "total de filas en revisión"
    ],
    "cantidad": [
        total_original,
        conteo_ph_ausente,
        conteo_ph_fuera_rango,
        conteo_fecha_invalida,
        total_evaluado 
    ]
})

controles.to_csv(OUT / "controles_agrosavia.csv", index=False)

display(controles)
print(OUT / "controles_agrosavia.csv")

,control,cantidad
0,total de entrada,92738
1,pH sin número,27213
2,pH fuera del intervalo,0
3,fecha no interpretable,0
4,total de filas en revisión,92738


OUT/controles_agrosavia.csv


**Interpretación del resultado:**

_Se construyo la tabla "controles" con las columnas control y cantidad, donde aparecen las filas de total de entrada, pH sin número, pH fuera del intervalo, fecha no interpretable y total de filas en revisión. Luego se hizo la exportacion a OUT / "controles_agrosavia.csv", donde seguidamente se comprobo su creación.Los hallazgos que se encontraron fueron que el numero de entradas es el mismo del total de filas en revision y otro hallazgo fue que el pH sin numero es de 27213, mientras que no hay valores en "pH fuera del intervalo" ni en "fecha no interpretable"._

## Entrega

El notebook completado se conserva como `soluciones/PD01_exploracion_agrosavia.ipynb`. La salida regenerable es `kit/salidas/soluciones/PD01/controles_agrosavia.csv`.

✅ Deben quedar visibles el perfil, los controles de conversión, la conciliación de filas y dos hallazgos. El notebook debe ejecutar todas las celdas en orden después de reiniciar el kernel.

⚠️ El intervalo de pH es un control de dominio para esta práctica. No determina por sí solo una recomendación agronómica. La frecuencia de análisis de suelo no representa automáticamente la distribución de los suelos de todo el territorio.